p2p/peer.go

```diff
@@ -901,117 +980,219 @@ func (m *PeerManager) SetHeight(peerID NodeID, height int64) error {
+// from disk on initialization, and any changes are written back to disk
+// (without fsync, since we can afford to lose recent writes).
 type peerStore struct {
-	peers map[NodeID]peerInfo
+	db     dbm.DB
+	peers  map[NodeID]*peerInfo
+	ranked []*peerInfo // cache for Ranked(), nil invalidates cache
+}
+
+// newPeerStore creates a new peer store, loading all persisted peers from the
+// database into memory.
+func newPeerStore(db dbm.DB) (*peerStore, error) {
+	store := &peerStore{
+		db: db,
+	}
+	if err := store.loadPeers(); err != nil {
+		return nil, err
+	}
+	return store, nil
 }
 
-// newPeerStore creates a new peer store.
-func newPeerStore() *peerStore {
-	return &peerStore{
-		peers: map[NodeID]peerInfo{},
+// loadPeers loads all peers from the database into memory.
+func (s *peerStore) loadPeers() error {
+	peers := make(map[NodeID]*peerInfo)
+
+	start, end := keyPeerInfoRange()
+	iter, err := s.db.Iterator(start, end)
+	if err != nil {
+		return err
+	}
+	defer iter.Close()
+	for ; iter.Valid(); iter.Next() {
+		// FIXME: We may want to tolerate failures here, by simply logging
+		// the errors and ignoring the faulty peer entries.
+		msg := new(p2pproto.PeerInfo)
+		if err := proto.Unmarshal(iter.Value(), msg); err != nil {
+			return fmt.Errorf("invalid peer Protobuf data: %w", err)
+		}
+		peer, err := peerInfoFromProto(msg)
+		if err != nil {
+			return fmt.Errorf("invalid peer data: %w", err)
+		}
+		peers[peer.ID] = peer
 	}
+	if iter.Error() != nil {
+		return iter.Error()
+	}
+	s.peers = peers
+	s.ranked = nil // invalidate cache if populated
+	return nil
 }
 
-// Get fetches a peer, returning nil if not found.
-func (s *peerStore) Get(id NodeID) (*peerInfo, error) {
+// Get fetches a peer. The boolean indicates whether the peer existed or not.
+// The returned peer info is a copy, and can be mutated at will.
+func (s *peerStore) Get(id NodeID) (peerInfo, bool) {
 	peer, ok := s.peers[id]
-	if !ok {
-		return nil, nil
+	return peer.Copy(), ok
+}
+
+// Set stores peer data. The input data will be copied, and can safely be reused
+// by the caller.
+func (s *peerStore) Set(peer peerInfo) error {
+	if err := peer.Validate(); err != nil {
+		return err
 	}
-	return &peer, nil
+	peer = peer.Copy()
+
+	if current, ok := s.peers[peer.ID]; !ok || current.Score() != peer.Score() {
+		// If the peer is new, or its score changes, we invalidate the Ranked() cache.
+		s.peers[peer.ID] = &peer
+		s.ranked = nil
+	} else {
+		// Otherwise, since s.ranked contains pointers to the old data and we
+		// want those pointers to remain valid with the new data, we have to
+		// update the existing pointer address.
+		*current = peer
+	}
+
+	// FIXME: We may want to optimize this by avoiding saving to the database
+	// if there haven't been any changes to persisted fields.
+	bz, err := peer.ToProto().Marshal()
+	if err != nil {
+		return err
+	}
+	return s.db.Set(keyPeerInfo(peer.ID), bz)
```

Comment:
So this isn't atomic wrt to the in-mem map. Why not store in the DB first and only upon no error, update `s.peers`.

Reply:
Good call, thanks.